src/main/scala/scalismo/registration/ImageMetric.scala

```diff
@@ -0,0 +1,120 @@
+
+  override val ndSpace = implicitly[NDSpace[D]]
+
+  def value(fixedImage: ScalarImage[D], movingImage: ScalarImage[D], transform: Transformation[D]) = {
+    computeValue(fixedImage, movingImage, transform, Integrator(sampler))
+  }
+
+  // compute the derivative of the cost function
+  def takeDerivativeWRTToTransform(fixedImage: ScalarImage[D], movingImage: DifferentiableScalarImage[D],
+    transformationSpace: TransformationSpace[D], params: DenseVector[Double]): DenseVector[Double] = {
+    computeDerivative(fixedImage, movingImage, transformationSpace, params, Integrator(sampler))
+  }
+
+  override def computeValueAndDerivative(fixedImage: ScalarImage[D], movingImage: DifferentiableScalarImage[D],
+    transformationSpace: TransformationSpace[D], params: DenseVector[Double]): ValueAndDerivative = {
+
+    // We create a new sampler, which always returns the same points. In this way we can make sure that the
+    // same sample points are used for computing the value and the derivative
+    val sampleOnceSampler = new Sampler[D] {
+      override val numberOfPoints: Int = sampler.numberOfPoints
+      private val samples = sampler.sample()
+
+      override def sample()(implicit rand: Random): IndexedSeq[(Point[D], Double)] = samples
+      override def volumeOfSampleRegion: Double = sampler.volumeOfSampleRegion
+    }
+
+    val integrator = Integrator(sampleOnceSampler)
+    val value = computeValue(fixedImage, movingImage, transformationSpace.transformForParameters(params), integrator)
+    val derivative = computeDerivative(fixedImage, movingImage, transformationSpace, params, integrator)
+    ValueAndDerivative(value, derivative)
+
+  }
+
+  private def computeValue(fixedImage: ScalarImage[D], movingImage: ScalarImage[D], transform: Transformation[D], integrator: Integrator[D]) = {
+    val warpedImage = movingImage.compose(transform)
+    def square(img: ScalarImage[D]) = img :* img
+
+    integrator.integrateScalar(square(fixedImage - warpedImage)) / integrator.sampler.volumeOfSampleRegion
+  }
+
+  private def computeDerivative(fixedImage: ScalarImage[D], movingImage: DifferentiableScalarImage[D],
+    transformationSpace: TransformationSpace[D], params: DenseVector[Double],
+    integrator: Integrator[D]): DenseVector[Double] = {
+
+    val transform = transformationSpace.transformForParameters(params)
+
+    val movingImageGradient = movingImage.differentiate
+    val warpedImage = movingImage.compose(transform) // compute the derivative of the cost function
+    val dDMovingImage = (fixedImage - warpedImage) * (-2.0 / sampler.volumeOfSampleRegion)
+
+    val dTransformSpaceDAlpha = transformationSpace.takeDerivativeWRTParameters(params)
+
+    val fullMetricGradient = (x: Point[D]) => {
+      val domain = Domain.intersection(fixedImage.domain, dDMovingImage.domain)
+      if (domain.isDefinedAt(x))
+        Some(dTransformSpaceDAlpha(x).t * (movingImageGradient(transform(x)) * dDMovingImage(x).toDouble).toBreezeVector)
```

Comment:
One advantage of the previous interface was that an eventual metric developer would only have to worry about differentiating the expression of the metric itself, the rest of the chain rule being taken care of by the Registration code. In this particular case, one would have to pay attention to to include the dTransformSpaceDAlpha.